### ISSUE TITLE
Request for Enhanced Type Annotations for JSX Elements

### ISSUE THREAD

**Issue by PeronGH** [honojs/hono#1409]
#### Overview

The current implementation of Hono JSX offers only minimal type annotations for JSX elements. This limitation impedes the developer experience by offering no code completion or error checking capabilities related to element properties (props). This document serves as a formal request to introduce comprehensive type annotations for JSX elements in Hono JSX to enrich both code quality and developer productivity.

#### Rationale

The lack of extensive type annotations hinders developers from leveraging the full capabilities of modern IDEs. This deficit could lead to:
- Missed opportunities for catching bugs at compile-time
- Reduced code maintainability due to a lack of clarity and code insights
- Hindered productivity as developers cannot make use of code completions or other intelligent IDE features.

#### Proposed Solution

Integrate exhaustive type annotations for element props, conforming to established best practices in the TypeScript community. This should involve defining interfaces or types that encapsulate the expected prop structures for each JSX element.

#### Expected Benefits

1. Improved Code Reliability: Comprehensive type checking will help in identifying issues early in the development cycle.
2. Enhanced Developer Experience: Code completion suggestions will speed up development and reduce context-switching.
3. Greater Maintainability: Explicit types will make the code more readable and easier to refactor.

**Comment by yusukebe** [comment]
Thanks, I'll comment later.

cc: @yudai-nkt

**Comment by yudai-nkt** [comment]
I had exactly the same motivation and you might want to take a look at https://github.com/yudai-nkt/hono-jsx-types. It seem you like Deno but I'm afraid my repo won't work with Deno out of the box because it depends on npm lifecycle script.

**Comment by yusukebe** [comment]
Hi @yudai-nkt 

Is it possible to create type definitions without relying on external npm libraries?

**Comment by yudai-nkt** [comment]
If we manually write down all the attributes for every tag in the type definition, then yes but that's definitely not the path we should go as you know.

**Comment by yusukebe** [comment]
Hmm. I'd like to enable this, but... Hmm.

**Comment by sjc5** [comment]
How is this handled in Sonik @yusukebe? Is that still using Preact, etc., for JSX rendering instead of Hono JSX? Would really love to have types for intrinsic elements built into Hono JSX for Hwy as well to keep Hono as the only dependency.

**Comment by yusukebe** [comment]
@sjc5, Sonik includes presets for hono-jsx, Preact, and React for rendering. However, similar to the this situation, hono-jsx is unable to handle the types of HTML elements.

**Comment by yusukebe** [comment]
@yudai-nkt,

I believe we can manage this by including the type definitions in the Hono core or another package. For instance, Preact has done it as seen here:

https://github.com/preactjs/preact/blob/main/src/jsx.d.ts

This file totals around 2,700 lines.

**Comment by sjc5** [comment]
Thanks @yusukebe, I would love to see it built in like Preact.

Meanwhile, this actually isn't too bad of a workaround.

1. Install Preact as development dependency only

```sh
npm i -D preact
```

2. Add a global override file

```ts
// jsx-override.d.ts

import type { JSX as PreactJSX } from "preact";

declare global {
  namespace JSX {
    interface IntrinsicElements extends PreactJSX.IntrinsicElements {}
  }
}
```

Perhaps there are downsides I haven't noticed yet, but on a quick test it seems to work nicely.

Perhaps Hono itself could even do this in its own JSX declaration (with Preact as a development dependency only) so it can build its own JSX declarations based on Preact's at the Hono build time?

**Comment by sjc5** [comment]
Hmm, on further inspection, the issue in my edit above (namely, that standard html like `onclick="run(arg);"` gives a type error) is also true using the https://github.com/yudai-nkt/hono-jsx-types approach as well.

I think this behavior isn't correct for most apps that use Hono JSX for server-rendering, because they are most likely using JSX just a convenient way to write escaped html. In that scenario, you want to add traditional html attributes (`onclick="run(arg);"`), not React-style handlers (`onClick={() => run(arg)}`), because apps using Hono JSX don't have the hydration step to actually attach those handlers (otherwise they would be using the applicable front-end framework's `renderToString()` equivalent instead).

So, to summarize, I think whatever solution is done here (whether through Hono or a third-party package) should support traditional html attributes, and not React-style handlers. Any apps using Hono to server-render an actual React, Preact, etc. app will be using the applicable `renderToString()` function provided by the front-end framework. Whereas apps using Hono JSX are purely server-side (not hydrated) and should support traditional html attributes.

**Comment by yusukebe** [comment]
Perhaps, we can use the DOM lib with a triple slash directive without writing it in tsconfig.json.

**Comment by yusukebe** [comment]
Or we can define them manually like React.

**Comment by sjc5** [comment]
This is very ugly, but I think that @yudai-nkt's solution with a little bit of extra type munging like this may work:

```ts
type BooleanLike = boolean | `${boolean}`;

type NumberLike = number | `${number}`;

type IfValueTypeIsBooleanConvertToBooleanOrBooleanString<T> = T extends boolean
  ? BooleanLike
  : T;

type IfValueTypeIsNumberConvertToNumberOrNumberString<T> = T extends number
  ? NumberLike
  : T;

type IfValueTypeIsFunctionConvertToString<T> = T extends (...args: any[]) => any
  ? string
  : T;

type HandleConversions<T> = IfValueTypeIsBooleanConvertToBooleanOrBooleanString<
  IfValueTypeIsNumberConvertToNumberOrNumberString<
    IfValueTypeIsFunctionConvertToString<T>
  >
>;
```

When the above is added into the HTMLAttributes definition in https://github.com/yudai-nkt/hono-jsx-types/blob/main/index.d.ts, it seems to have good results.

Still experimenting though.

**Comment by yusukebe** [comment]
Hey @PeronGH @yudai-nkt @sjc5 !

cc: @usualoma @Code-Hex @sor4chi 

I've made the PR #1614 to resolve this issue. Please take a look!

**Comment by Code-Hex** [comment]
@yusukebe Thanks to mention me!

In my personal opinion, I have mixed feelings about it. Currently, in my project, JSX behaves as a simple HTML template. I'm using JSX, but I'm not using React. So, with this autocomplete feature, for example, it might work for "className," but in hono/jsx, you don't write "class" to use it, so there might be a difference between the autocomplete suggestions and the rendering.

Am I overthinking this?

**Comment by yusukebe** [comment]
Hi @Code-Hex 

The point is that the type definitions are optimized for hono/jsx (standard-HTML base) while being ported from React. So, `className` is not used; it's `class`. And `htmlFor` is not used; use `for` instead.

<img width="634" alt="Screenshot 2023-10-24 at 0 06 42" src="https://github.com/honojs/hono/assets/10682/605134c3-bb26-4998-8396-3ec56634ac49">

<img width="704" alt="Screenshot 2023-10-24 at 0 08 04" src="https://github.com/honojs/hono/assets/10682/fb7fb56c-aa1a-473c-8d5e-396b89b393b8">

I believe this meets your use case.

**Comment by sjc5** [comment]
Thank you so much for your time and attention on this @yusukebe! 🙏 

I haven't tested yet, but it looks like this is missing the base event handler attributes (`onclick`, etc.), though it includes the tag-specific ones like `onchange`. Though it should be pretty easy to add in a base list of common event attributes.

I'm not sure which way is better (writing everything manually, and perhaps a little too narrowly) or starting with the broad TypeScript dom types and narrowing them down (resulting in perhaps a too broad set, and with more complicated types).

I will give this a try as soon as I can, but in the meantime, here is an example of where I landed with my experimentation last night in case anyone is curious: https://github.com/hwy-js/hwy/blob/c230da30efff4b6c46376ce08ee8393722508b87/packages/dev/index.ts. In the end I think your approach is probably better.

**Comment by yusukebe** [comment]
Hi @sjc5,

Yes, your approach is another one I considered to resolve this issue. I didn't follow that path because I didn't want to create a dependency on the `DOM` lib. I'm not sure to what extent depending on the `DOM` lib would affect a Hono app using JSX. So if it's okay to depend on the DOM `lib`, then your way might not be a bad choice.

Regardless, I believe the current approach is better.

**Comment by usualoma** [comment]
Hi @yusukebe!

#1614 is an awesome PR.

It would be even better to put the declaration in the Hono namespace so that users can override it in their applications. #1616

If the user adds the following to the application...

```ts
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Hono {
    interface HTMLAttributes {
      'data-custom-attr'?: 'foo' | 'bar';
    }
    interface IntrinsicElements {
      'my-custom-element': Hono.HTMLAttributes & {
        'x-event'?: 'click' | 'scroll';
      }
    }
  }
}
```

Type checking is enabled as follows.

<img width="585" alt="image" src="https://github.com/honojs/hono/assets/30598/0a2302dd-86ca-44f0-bbfb-b41a48034624">

<img width="682" alt="image" src="https://github.com/honojs/hono/assets/30598/0b3d0445-3865-453d-aeee-9dab58e1c9fd">

**Comment by yusukebe** [comment]
@usualoma 

Looks great to me! I'll merge it into #1614 later!

**Comment by yusukebe** [comment]
This can be closed! Thanks!
